packages/zoe/src/contractFacet/vatRoot.js

```diff
@@ -8,51 +8,84 @@
 // `yarn build-zcfBundle`.
 
 import { Far } from '@endo/marshal';
+import { E } from '@endo/far';
 
 import '../../exported.js';
 import '../internal-types.js';
 
 import { makeZCFZygote } from './zcfZygote.js';
 
+const { details: X } = assert;
+
 /**
  * @param {VatPowers & { testJigSetter: TestJigSetter }} powers
- * @returns {{ executeContract: ExecuteContract}}
+ * @param {{contractBundleCap: BundleCap, zoeService: ZoeService, invitationIssuer: Issuer, privateArgs: any}} vatParameters
+ * @param {import('@agoric/vat-data').Baggage} baggage
  */
-export function buildRootObject(powers) {
+// * @returns {{ executeContract: ExecuteContract}}
+export async function buildRootObject(powers, vatParameters, baggage) {
   // Currently, there is only one function, `executeContract` called
   // by the Zoe Service. However, when there is kernel support for
   // zygote vats (essentially freezing and then creating copies of
   // vats), `makeZCFZygote`, `zcfZygote.evaluateContract` and
   // `zcfZygote.startContract` should exposed separately.
   const { testJigSetter } = powers;
+  const { contractBundleCap } = vatParameters;
+  assert(
+    contractBundleCap !== undefined,
+    X`expected vatParameters.contractBundleCap ${vatParameters}`,
+  );
+  let { zoeService, invitationIssuer } = vatParameters;
+  const didStart = baggage.has('DidStart');
+  if (!didStart) {
+    baggage.init('DidStart', true);
+    baggage.init('zoeService', zoeService);
+    baggage.init('invitationIssuer', invitationIssuer);
+  } else {
+    assert(!zoeService);
+    zoeService = baggage.get('zoeService');
+
+    assert(!invitationIssuer);
+    invitationIssuer = baggage.get('invitationIssuer');
+  }
 
-  /** @type {ExecuteContract} */
-  const executeContract = (
-    bundleOrBundleCap,
+  // make zcfZygote with contract-general state and kinds initialized
+  const zcfZygote = await makeZCFZygote(
+    powers,
     zoeService,
     invitationIssuer,
-    zoeInstanceAdmin,
-    instanceRecordFromZoe,
-    issuerStorageFromZoe,
-    privateArgs = undefined,
-  ) => {
-    /** @type {ZCFZygote} */
-    const zcfZygote = makeZCFZygote(
-      powers,
-      zoeService,
-      invitationIssuer,
-      testJigSetter,
+    testJigSetter,
+    contractBundleCap,
+    baggage,
+  );
+
+  // snapshot zygote here //////////////////
+
+  if (didStart) {
+    return E.when(E(zcfZygote).restartContract(vatParameters.privateArgs), () =>
+      Far('upgraded contractRunner', {}),
     );
-    zcfZygote.evaluateContract(bundleOrBundleCap);
-    return zcfZygote.startContract(
+  }
+
+  return Far('contractRunner', {
+    // initialize instance-specific state of the contract
+    startZcf: (
       zoeInstanceAdmin,
       instanceRecordFromZoe,
       issuerStorageFromZoe,
-      privateArgs,
-    );
-  };
+      privateArgs = undefined,
```

Comment:
Ah, I didn't see that annotation in the view.

Nothing to be repeated here. I think I'm suggesting removing `= undefined` because it doesn't change the runtime and the documentation value is already in the jsdoc

Reply:
> Where should I put this explanation?

[Types documentation](https://github.com/Agoric/agoric-sdk/wiki/Types-documentation), part of [Coding Style](https://github.com/Agoric/agoric-sdk/wiki/Coding-Style)

though keep in mind that the wiki genre works best for capturing a consensus view and less so for arguing a position. So you could open an issue or use some other discussion tool. Or you could make an executive decision.

Reply:
https://github.com/Agoric/agoric-sdk/wiki/Express-optional-parameters-in-both-JS-and-JSDoc-TS-syntax

Reply:
> The optionality of a function is an essential part of its non-formal meaning.

Fully agree.

> Types are often textually distant.

Are you saying this is too distant?
```js
 * @param {string] [optionalStr]
 */
const fn = (optionalStr) => {
```

I recognize functions often have more params, but optional ones are always at the end of the list.

> JS has its own syntax for optional parameters.

The only syntax I know in JS for expressing optional parameters is JSDoc, which I'm advocating. What reference is there that ` = undefined` is the Javascript standard way of expression optional parameters?

> We should use that.

This discussion isn't just stylistic. Using ` = foo` is a type hint for the domain of the value. If we write `= undefined` then when there's no other hint that constrains the domain from `any` the literal `undefined`. This is why you had to write `non-null" four times in https://github.com/Agoric/agoric-sdk/blob/6fafffc64e13a9f6e5029c1534bce6c64454762d/packages/zoe/src/contractFacet/zcfMint.js#L76

I think this is worth further discussion before canonizing in the wiki.